File: src/MauiTemplatesCLI/MauiAppX/MauiProgram.cs
#if Razor
using BlazorBindings.Maui;
#endif
#if (AddToolkitPackage || AddMediaPackage)
using CommunityToolkit.Maui;
#endif
#if Hybrid
using MauiApp._1.Data;
#endif
#if Net7OrLater
using Microsoft.Extensions.Logging;
#endif
#if AddFoldablePackage
using Microsoft.Maui.Foldable;
#endif
#if (AddToolkitPackage || Hybrid || Net7OrLater || AddFoldablePackage || Razor)

#endif
namespace MauiApp._1
{
    public static partial class MauiProgram
    {
        public static MauiApp CreateMauiApp()
        {
            var builder = MauiApp.CreateBuilder();
            builder.UseMauiApp<App>()
#if Razor
                   .UseMauiBlazorBindings()
#endif
#if AddFoldablePackage
                   .UseFoldable()
#endif
#if AddMapsPackage
                   .UseMauiMaps()
#endif
#if AddToolkitPackage
                   .UseMauiCommunityToolkit()
#endif
#if AddMarkupPackage
                   .UseMauiCommunityToolkitMarkup()
#endif
#if AddMediaPackage
                   .UseMauiCommunityToolkitMediaElement()
#endif
                   .ConfigureFonts(fonts =>
                   {
                       fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
                       fonts.AddFont("OpenSans-SemiBold.ttf", "OpenSansSemiBold");
                   });

#if (Mvvm && !(Hybrid || Razor))
#if (Plain || Markup)
            builder.Services.AddSingleton(SemanticScreenReader.Default);
            builder.Services.AddSingleton<MainViewModel>();
            builder.Services.AddSingleton<MainPage>();

#else
            builder.Services.AddSingleton<IDialogService, DialogService>();
            builder.Services.AddSingleton<INavigationService, NavigationService>();
#endif
#if Hierarchical
            builder.Services.AddSingleton<MainViewModel>();
            builder.Services.AddSingleton<MainPage>();
            builder.Services.AddTransient<NewEventViewModel>();
            builder.Services.AddTransient<NewEventPage>();

#endif
#if Tabbed
            builder.Services.AddSingleton<EventsViewModel>();
            builder.Services.AddSingleton<SearchViewModel>();
            builder.Services.AddSingleton<SettingsViewModel>();
            builder.Services.AddTransient<NewEventViewModel>();
            builder.Services.AddTransient<NewEventPage>();

#endif
#if Shell
            builder.Services.AddSingleton<EventsViewModel>();
            builder.Services.AddSingleton<EventsPage>();
            builder.Services.AddSingleton<SearchViewModel>();
            builder.Services.AddSingleton<SearchPage>();
            builder.Services.AddSingleton<SettingsViewModel>();
            builder.Services.AddSingleton<SettingsPage>();
            builder.Services.AddTransient<LoginViewModel>();
            builder.Services.AddTransient<LoginPage>();
            builder.Services.AddTransient<NewEventViewModel>();
            builder.Services.AddTransient<NewEventPage>();

#endif
#endif
#if Hybrid
            builder.Services.AddMauiBlazorWebView();
            // Caution: Recommended to enable Developer Tools only for development!!!
#if Net7OrLater
//-:cnd:noEmit
#if DEBUG
            builder.Services.AddBlazorWebViewDeveloperTools();
            builder.Logging.AddDebug();
#endif
//+:cnd:noEmit
#else
//-:cnd:noEmit
#if DEBUG
            builder.Services.AddBlazorWebViewDeveloperTools();
#endif
//+:cnd:noEmit
#endif
            builder.Services.AddSingleton(AppInfo.Current);
            builder.Services.AddSingleton<WeatherForecastService>();
#elif Net7OrLater
//-:cnd:noEmit
#if DEBUG
            builder.Logging.AddDebug();
#endif
//+:cnd:noEmit
#endif

            return builder.Build();
        }
    }
}

File: src/MauiTemplatesCLI/MauiAppFS/MauiProgram.fs
namespace MauiApp._1

open Microsoft.Extensions.DependencyInjection
open Microsoft.Maui.Accessibility
open Microsoft.Maui.Hosting
open Fabulous.Maui
#if AddToolkit
open CommunityToolkit.Maui
#endif
#if AddSyncfusionToolkit
open Syncfusion.Maui.Toolkit.Hosting
#endif
#if AddMarkup
open CommunityToolkit.Maui.Markup
#endif
#if AddFoldable
open Microsoft.Maui.Foldable
#endif
#if AddMaps
open Fabulous.Maui.Maps
#if (AllPlatforms || IsWindows)
open CommunityToolkit.Maui.Maps
#endif
#endif
#if AddMedia
open Fabulous.Maui.MediaElement
#endif
open MauiApp._1.Extensions
#if Hybrid
open MauiApp._1.RazorLib.Data
#endif

type MauiProgram =
    static member CreateMauiApp() =
        MauiApp
            .CreateBuilder()
            .UseFabulousApp(App.program)
#if AddToolkit
            .UseMauiCommunityToolkit()
#endif
#if AddSyncfusionToolkit
            .ConfigureSyncfusionToolkit()
#endif
#if AddMarkup
            .UseMauiCommunityToolkitMarkup()
#endif
#if AddFoldable
            .UseFoldable()
#endif
#if AddMaps
            .UseFabulousMaps()
#if (AllPlatforms || IsWindows)
            .UseMauiCommunityToolkitMaps("<BING_MAPS_API_KEY_HERE>") // To generate a Bing Maps API Key, visit https://www.bingmapsportal.com/
#endif
#endif
#if AddMedia
            .UseFabulousMediaElement()
#endif
#if Hybrid
            .ConfigureBlazorWebView()
#endif
            .ConfigureFonts(fun fonts ->
                fonts
                    .AddFont("OpenSans-Regular.ttf", "OpenSansRegular")
                    .AddFont("OpenSans-Semibold.ttf", "OpenSansSemibold")
                |> ignore)
            .ConfigureServices(fun services ->
                services
#if Hybrid
                    .AddSingleton<WeatherForecastService>()
//-:cnd:noEmit
#if DEBUG
                    .AddBlazorWebViewDeveloperTools()
#endif
//+:cnd:noEmit
#else
                    .AddSingleton(SemanticScreenReader.Default)
#endif
                |> ignore)
//-:cnd:noEmit
#if DEBUG
            .AddDebugLog()
#endif
//+:cnd:noEmit
            .Build()
